### PR TITLE
Pass simulator into CameraDevice when setting up for tests.

### PR DESCRIPTION
**Pull request by Brow71189**
This is needed for setting up the tests after the changes to nionswift-usim submitted in https://github.com/nion-software/nionswift-usim/pull/34.

### CODE REVIEW

**Comment by cmeyer** [comment]
I'd like to consider doing this a different way... I think encapsulation is broken in the current code and this change just further erodes encapsulation.

My proposal:

- do not track cameras in the [simulator] instrument device
- track the simulator in the camera device by adding a field `__simulator` and read only property `simulator`
- eliminate `__cameras` in the simulator instrument device
- replace `_get_camera_simulator` calls with `camera_device.simulator` or similar
- replace calls to `get_camera_data` with `camera_device.get_camera_data` or similar
- eliminate `get_camera_dimensional_calibrations`
- change `parse_camera_values` to ask the camera device for dimensional calibrations
- use `Registry` for lookup by camera type (I think this will work, might need some tweaking to get it right)

**Comment by Brow71189** [comment]
Ok I'll make the changes. But this pull request here will still be needed as is, right? Or did you not want to pass the simulator into the camera device? Btw, it is already being tracked in the camera device (there is a field ``__camera_simulator``), the additional "registration" in instrument was just there for backwards-compatibility reasons.

**Comment by Brow71189** [comment]
> 
>     * use `Registry` for lookup by camera type (I think this will work, might need some tweaking to get it right)

How do you mean that? Should we register camera devices under their type instead of the generic ``camera_module`` used now? So would we just call it ``ronchigram_camera_device`` or something like that in the registry?

**Comment by cmeyer** [comment]
> 

The simulator can be created inside the camera device init. The overall goal is to reduce/eliminate usim internal/implementation details from instrumentation kit tests.

File: nion/instrumentation/test/AcquisitionTestContext.py [hunk 1]
```diff
@@ -76,13 +76,9 @@ def setup_camera_hardware_source(self, stem_controller: stem_controller.STEMCont
...
-        if camera_type == "ronchigram":
-            camera_simulator = RonchigramCameraSimulator.RonchigramCameraSimulator(instrument, Geometry.IntSize.make(instrument.camera_sensor_dimensions("ronchigram")), instrument.counts_per_electron, instrument.stage_size_nm)
-        else:
-            camera_simulator = EELSCameraSimulator.EELSCameraSimulator(instrument, Geometry.IntSize.make(instrument.camera_sensor_dimensions("eels")), instrument.counts_per_electron)
-        camera_device = CameraDevice.Camera(camera_id, camera_type, camera_name, instrument, camera_simulator)
-        if hasattr(instrument, "_set_camera_simulator"):
-            instrument._set_camera_simulator(camera_type, camera_simulator)
+        camera_device = CameraDevice.Camera(camera_id, camera_type, camera_name, instrument)
+        # Also register the camera device and use a unique name for it so that we can directly access it
+        Registry.register_component(camera_device, {f"usim_{camera_type}_camera_device"})
```
[the remark this answers is not in the document]
Comment (Brow71189) [reply nion/instrumentation/test/AcquisitionTestContext.py:81]: This is used in ``parse_camera_values" in here: https://github.com/Brow71189/nionswift-usim/blob/deb91863af1301542d0d2f70eda5055ddd13741b/nionswift_plugin/usim/InstrumentDevice.py#L780

[the remark this answers is not in the document]
Comment (Brow71189) [reply nion/instrumentation/test/AcquisitionTestContext.py:81]: Isn't moving the ``register_component`` call to ``CameraDevice.__init__``  a bit dangerous? I'm thinking that if you create two instances of ``CameraDevice``  with the same camera type, the ``register_component`` call will fail. Or should we just check if the device is already registered and only register it then? At least that would make the patch to ``instrumentation-kit``  obsolete.

[the remark this answers is not in the document]
Comment (cmeyer) [reply nion/instrumentation/test/AcquisitionTestContext.py:81]: Discussed offline: assert that the component isn't already registered. Tests should be very strict about deregistering anything they register and it shouldn't come up in non-tests ever.

**Comment by Brow71189** [comment]
Not needed anymore with latest commit in the usim PR.

**Pull request closed without merging**